File: medium/palindromic-substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [[0] *n ] * n
        
        count = 0
        # bottom up
        
        
        def recurse_odd(ind):
            nonlocal count 
            l, r = ind, ind #odd case
            
            while 0 <= l < n and 0 <=  r < n and s[l] == s[r]:
                count += 1

                l -=1
                r += 1
       
    
        def recurse_even(ind):
            nonlocal count 
            l, r = ind, ind +1 #odd case
            
            while 0 <= l < n and 0 <=  r < n and s[l] == s[r]:
                count += 1

                l -=1
                r += 1
                   
        for i in range(n):
            recurse_odd(i)
            if i != n-1:
                recurse_even(i)
        
            
        return count

File: medium/palindromic-substrings.cpp
// A hash function used to hash a pair of any kind
struct hash_pair {
    size_t operator()(const pair<int, int>& p) const
    {
        auto hash1 = hash<int>{}(p.first);
        auto hash2 = hash<int>{}(p.second);
 
        if (hash1 != hash2) {
            return hash1 ^ hash2;             
        }
         
        // If hash1 == hash2, their XOR is zero.
          return hash1;
    }
};

class Solution {
public:
    int countSubstrings(string s) {
        
        int n = s.size();
        
        vector<vector<bool>> dp;
        
        for (int i =0 ; i<n; i++){
            vector<bool> row;
            for (int j =0; j<n; j++){
                row.push_back(false);
            }
            dp.push_back(row);
        }    
        int count = 0;
        for (int i =0; i < n; ++i){
            if (i < n- 1){
                dp[i][i +1] = s[i] == s[i+1];
                if (dp[i][i +1]) count ++;
            }
            dp[i][i] = true; 
            count ++;
        }
        
        for (int k = 3; k <= n; ++k){
            for (int i=0; i <= n - k ; ++i){
                int j = i + k -1;
                dp[i][j] = dp[i+1][j-1] && s[i] == s[j];
                if (dp[i][j]) count ++;
            }
        }
        return count; 
    }
};
